### PR TITLE
chore(nns): NeuronsFundAuditInfo migration

### PR DESCRIPTION
**Pull request by anchpop**
## Problem

We have switched from the `hotkey_principal` field to a field named `nns_neuron_controller` in the type `NeuronsFundNeuronPortion`. This is because the type `nns_neuron_controller` more clearly explains the semantics of the type than `hotkey_principal`. However, already-existing `ProposalData`s indirectly contain `NeuronsFundNeuronPortion`s which have `None` for the new field.

## Solution

I've implemented a migration for the `NeuronsFundNeuronPortion`s that are transitively stored in `ProposalData` to use this new field. The migration lives in `pre_upgrade` and iterates over the stored `ProposalData`s, and sets their `neurons_fund_data` field to a value that contains the migrated state. 

< [Previous PR](https://github.com/dfinity/ic/pull/1171) |

### CODE REVIEW

File: rs/nns/governance/canister/canister.rs [hunk 1]
```diff
@@ -395,6 +396,17 @@ fn canister_post_upgrade() {
     validate_stable_storage();
 }
 
+fn migrate_neurons_fund_audit_info(state: InternalGovernanceProto) -> InternalGovernanceProto {
+    InternalGovernanceProto {
+        proposals: state
+            .proposals
+            .into_iter()
+            .map(|(id, proposal_data)| (id, proposal_data.into_current()))
```
Comment (aterga) [inline rs/nns/governance/canister/canister.rs:404]: Nit. Use ::from?

File: rs/nns/governance/src/migrations.rs [hunk 2]
```diff
@@ -27,3 +32,93 @@ pub(crate) fn maybe_run_migrations(
     // TODO: move inactive neuron migration here.
     migrations
 }
+
+impl NeuronsFundData {
+    pub fn into_current(self) -> NeuronsFundData {
```
Comment (aterga) [inline rs/nns/governance/src/migrations.rs:37]:
```suggestion
    pub fn migrate(self) -> NeuronsFundData {
```

File: rs/nns/governance/src/migrations.rs [hunk 3]
```diff
@@ -27,3 +32,93 @@ pub(crate) fn maybe_run_migrations(
...
+    #[allow(deprecated)] // hotkey_principal is deprecated to make sure we don't accidentally use it.
+    pub fn into_current(self) -> NeuronsFundNeuronPortion {
+        NeuronsFundNeuronPortion {
+            nns_neuron_id: self.nns_neuron_id,
+            amount_icp_e8s: self.amount_icp_e8s,
+            maturity_equivalent_icp_e8s: self.maturity_equivalent_icp_e8s,
+            is_capped: self.is_capped,
+            controller: self.controller.or(self.hotkey_principal),
+            hotkeys: self.hotkeys,
+            hotkey_principal: self.hotkey_principal,
```
Comment (aterga) [inline rs/nns/governance/src/migrations.rs:121]:
```suggestion
        ..self
```

File: rs/nns/governance/src/migrations.rs [hunk 4]
```diff
@@ -27,3 +32,93 @@ pub(crate) fn maybe_run_migrations(
     // TODO: move inactive neuron migration here.
     migrations
 }
+
+impl NeuronsFundData {
```
Comment (aterga) [inline rs/nns/governance/src/migrations.rs:36]: Can we move this migration to a separate file?

File: rs/nns/governance/canister/canister.rs [hunk 5]
```diff
@@ -385,6 +385,7 @@ fn canister_post_upgrade() {
         restored_state.neurons.len(),
         restored_state.xdr_conversion_rate,
     );
+    let restored_state = migrate_neurons_fund_audit_info(restored_state);
```
Comment (aterga) [inline rs/nns/governance/canister/canister.rs:388]: Is this the right place to run the migration? Cc @max-dfinity

**Comment by anchpop** [comment]
```
//rs/nns/integration_tests:upgrade_canisters_with_golden_nns_state       PASSED in 1133.4s

Executed 1 out of 1 test: 1 test passes.
$ git status
On branch @anchpop/NeuronsFundAuditInfo-migration
```

**Comment by aterga** [comment]
> ```
> $ git status
> On branch @anchpop/NeuronsFundAuditInfo-migration
> 
> $ bazel test --test_env=SSH_AUTH_SOCK --test_env=NNS_CANISTER_UPGRADE_SEQUENCE=governance,root --test_output=streamed --test_arg=--nocapture //rs/nns/integration_tests:upgrade_canisters_with_golden_nns_state
> [...]
> //rs/nns/integration_tests:upgrade_canisters_with_golden_nns_state       PASSED in 1133.4s
> 
> Executed 1 out of 1 test: 1 test passes.
> ```

Thanks! Just to be sure the right code was executed, may I suggest adding a panic in the post_upgrade before you call the migration, and running the test once more? 

Nit: You don't need to also upgrade Root, so `NNS_CANISTER_UPGRADE_SEQUENCE=governance` should suffice.

File: rs/nns/integration_tests/BUILD.bazel [hunk 6]
```diff
@@ -417,3 +419,24 @@ rust_ic_test(
     ],
     deps = DEPENDENCIES + DEV_DEPENDENCIES,
 )
+
+rust_ic_test(
+    name = "hotkeys_migration_test",
```
Comment (max-dfinity) [inline rs/nns/integration_tests/BUILD.bazel:424]: This will be removed after deployment?

File: rs/nns/governance/canister/canister.rs (hunk 5 above)
Comment (max-dfinity) [reply rs/nns/governance/canister/canister.rs:388]: Yes, assuming the migration isn't very expensive.

File: rs/nns/integration_tests/BUILD.bazel (hunk 6 above)
Comment (anchpop) [reply rs/nns/integration_tests/BUILD.bazel]: This change has been moved to [this PR](https://github.com/dfinity/ic/pull/1171), but yes that is the current plan

File: rs/nns/governance/canister/canister.rs (hunk 5 above)
Comment (anchpop) [reply rs/nns/governance/canister/canister.rs:388]: 🙏

**Comment by anchpop** [comment]
> > ```
> > $ git status
> > On branch @anchpop/NeuronsFundAuditInfo-migration
> > 
> > $ bazel test --test_env=SSH_AUTH_SOCK --test_env=NNS_CANISTER_UPGRADE_SEQUENCE=governance,root --test_output=streamed --test_arg=--nocapture //rs/nns/integration_tests:upgrade_canisters_with_golden_nns_state
> > [...]
> > //rs/nns/integration_tests:upgrade_canisters_with_golden_nns_state       PASSED in 1133.4s
> > 
> > Executed 1 out of 1 test: 1 test passes.
> > ```
> 
> Thanks! Just to be sure the right code was executed, may I suggest adding a panic in the post_upgrade before you call the migration, and running the test once more?
> 
> Nit: You don't need to also upgrade Root, so `NNS_CANISTER_UPGRADE_SEQUENCE=governance` should suffice.

Doing this causes the test to fail as expected!

File: rs/nns/integration_tests/BUILD.bazel (hunk 6 above)
Comment (aterga) [reply rs/nns/integration_tests/BUILD.bazel]: Yes, I think we're going to remove it (or maybe convert to a test that still makes sense after the migration, if there's value to be extracted from it for free; let's discuss).

**Pull request merged**